sql/src/test/java/io/crate/metadata/rule/ingest/IngestionServiceTest.java

```diff
@@ -0,0 +1,151 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata.rule.ingest;
+
+import io.crate.metadata.Schemas;
+import io.crate.metadata.TableIdent;
+import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.test.ClusterServiceUtils;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class IngestionServiceTest extends CrateDummyClusterServiceUnitTest {
+
+    private final String UNDER_TEST_SOURCE_IDENT = "mqtt";
+    private final String UNDER_TEST_RULE_TARGET_TABLE = "temp_data.raw";
+    private final IngestRule UNDER_TEST_INGEST_RULE =
+        new IngestRule("temperature", UNDER_TEST_RULE_TARGET_TABLE, "(topic like 'temperature/%')");
+
+    private Schemas schemas;
+    private IngestRulesMetaData ingestRulesMetaData;
+    private IngestionService ingestionService;
+
+    @Before
+    public void setupIngestionServiceAndIngestRulesMetaData() {
+        schemas = mock(Schemas.class);
+        ingestionService = new IngestionService(schemas, clusterService);
+        ImmutableOpenMap.Builder<String, MetaData.Custom> customsBuilder = ImmutableOpenMap.builder();
+        ingestRulesMetaData = new IngestRulesMetaData(new HashMap<>());
+        customsBuilder.put(IngestRulesMetaData.TYPE, ingestRulesMetaData);
+        ImmutableOpenMap<String, MetaData.Custom> customs = customsBuilder.build();
+
+        MetaData metaData = MetaData.builder().customs(customs).build();
+        ClusterState clusterState = ClusterState.builder(clusterService.state())
+            .metaData(metaData)
+            .build();
+
+        ClusterServiceUtils.setState(clusterService, clusterState);
+        setupIngestRules();
+    }
+
+    private void setupIngestRules() {
+        ingestRulesMetaData.createIngestRule(UNDER_TEST_SOURCE_IDENT, UNDER_TEST_INGEST_RULE);
+        IngestRule humidityIngestRule = new IngestRule("humidity", "humidity_data.raw", "(topic like 'humidity/%')");
+        ingestRulesMetaData.createIngestRule("kafka", humidityIngestRule);
+    }
+
+    @Test
+    public void testAddListenerPushesCurrentRulesToListener() {
+        ingestionService.addListener(new IngestRuleListener() {
+            @Override
+            public String getSourceIdent() {
+                return UNDER_TEST_SOURCE_IDENT;
+            }
+
+            @Override
+            public void onRulesChanged(Set<IngestRule> rules) {
```

Comment:
currently addListener returns the current rules, so it'll be triggered once. Now that you mention it maybe it should be a separate method to initially get the rules?